CMSgov/easi-app
# Allow uploading/fetching/deleting documents for TRB forms

# EASI-2344

## Changes and Description

- Allow uploading, fetching, and deleting documents associated with a TRB request.
    - This uses the simpler upload process that MINT used (see [this MINT PR](https://github.com/CMSgov/mint-app/pull/220)); uploading documents are sent to our backend, then our backend uploads them to S3. When a user needs to view/download a document, our backend queries S3 to generate a presigned URL for accessing the document, then passes that to the frontend to relay to the user.
    - A new database table, `trb_request_documents`, tracks document metadata that's not stored in S3. We don't need to store the virus scan status of documents in our database, that's handled by tags in S3.
    - `pkg/graph/resolvers/trb_request_document.go` wires together the business logic for managing uploads, calling `s3Client`'s methods to interact with S3, and storage methods to interact with our database.
    - The keys for documents in S3 are _not_ the filenames; instead, they're UUIDs, with an extension (as a convenience for us if we needed to track down a document in S3 directly). The extension is pulled from the filename if present, with `.unknown` as a fallback. I didn't bother trying to track the content-type of uploads or set it in S3; it doesn't appear to matter from testing. If it does, it's fairly simple to add. (See the code removed in https://github.com/CMSgov/easi-app/pull/1819/commits/bcf2a75e9b2cc8002d6bd87c538b64d03749a75f)
- Configuring environment variables and related settings to make sure that the MinIO container can be connected to in all circumstances.
    - When running locally, both the `easi` backend container and the server tests use the `MINIO_ADDRESS` env var, which will be set by `.envrc` to `"http://host.docker.internal:9000"`. This address can be used both inside containers and from the host, due to the MinIO container publishing port 9000.
    - When running server tests in CI, the server tests will always run outside a container, so hardcoding `MINIO_ADDRESS: http://localhost:9000` in `.github/workflows/run_tests.yml` works to allow the tests to access MinIO.
    - When running e2e tests in CI, the backend container needs to be able to access the MinIO container, but we know all relevant code will be running on containers, so we can hardcode `MINIO_ADDRESS=http://minio:9000` in `docker-compose.cypress_ci.yml`.
- Adding a script `scripts/upload_trb_request_doc` that can be used to upload a document, adapted from the scripts in [GQLGen's upload documentation](https://gqlgen.com/reference/file-upload/). See below for more details on use.
- Adding a script `scripts/minio_cli` that starts up a container with the MinIO `mc` command-line client, connected to our docker compose network for local debugging.
- Small refactors of related code, such as pulling the key name `"av-status"` out into a constant.

## How to test this change

### Setup and uploading document

First, create a TRB request:
```gql
mutation {
  createTRBRequest(requestType: NEED_HELP) {
    id
  }
}
```
note the returned request ID.

Then, upload a document with `scripts/upload_trb_request_doc`. Replace the placeholder value for `request_id` with the actual request ID, then run the script from the root of the repo.

You can verify that the document is present either by checking the `./tmp/minio/easi-app-file-uploads/` directory within your local repo (which is mounted to the MinIO container) or by running `scripts/minio_cli`, then running `mc ls --recursive local` to list all files in MinIO. Note down the filename in MinIO/under the `./tmp` directory; this is used as the S3 key for the object in MinIO, which can be used with `scripts/tag_minio_file` to check virus checking.

### Testing fetching documents
GQL query to use:
```gql
query {
  trbRequest(id:"<trb request ID>") {
    documents {
      id
      fileName
      documentType {
        commonType
        otherTypeDescription
      }
      status
      uploadedAt
      url
    }
  }
}
```
`fileName` and `documentType` should always match what the doc was uploaded with, `uploadedAt` should always be the time when it was uploaded.

* On first fetch - `status` should be `"PENDING"`.
* Run `scripts/tag_minio_file "/easi-app-file-uploads/<S3 key>" CLEAN`, then rerun the GQL query - `status` should be `"AVAILABLE"`.
* Run `scripts/tag_minio_file "/easi-app-file-uploads/<S3 key>" INFECTED`, then rerun the GQL query - `status` should be `"UNAVAILABLE"`.

Testing URL:
* Run `scripts/minio_cli` to get a shell in a container attached to the local Docker Compose network.
* Inside that container, run `curl -o file.txt <presigned URL from GQL query, with "minio:9000" as the host>`.
* Run `cat file.txt`; it should output the uploaded file's contents.

### Testing deleting documents

Get the document's ID from the GQL query for trbRequest, then run:

```gql
mutation {
  deleteTRBRequestDocument(id:"<doc ID>") {
    document {
      fileName
    }
  }
}
```
This should delete the document, which can be confirmed by re-running the GQL query for trbRequest and/or checking the database. The file in MinIO will _not_ be deleted, however; the existing code for 508 documents similarly preserves the saved document.
